kms/tpmkms/tpmkms.go

```diff
@@ -326,9 +340,30 @@ func (k *TPMKMS) CreateKey(req *apiv1.CreateKeyRequest) (*apiv1.CreateKeyRespons
 		return nil, fmt.Errorf("failed parsing %q: %w", req.Name, err)
 	}
 
-	v, ok := signatureAlgorithmMapping[req.SignatureAlgorithm]
-	if !ok {
-		return nil, fmt.Errorf("TPMKMS does not support signature algorithm %q", req.SignatureAlgorithm)
+	ctx := context.Background()
+	caps, err := k.tpm.GetCapabilities(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("could not get TPM capabilities %w", err)
+	}
+
+	var v algorithmAttributes
+	var ok bool
+	if !properties.ak && req.SignatureAlgorithm == apiv1.UnspecifiedSignAlgorithm {
+		for _, alg := range preferredAlgorithm {
+			v, ok = signatureAlgorithmMapping[alg]
+			if caps.SupportsAlgorithms(v.Requires...) {
+				break
+			}
+		}
+	} else {
+		v, ok = signatureAlgorithmMapping[req.SignatureAlgorithm]
+		if !ok {
+			return nil, fmt.Errorf("TPMKMS does not support signature algorithm %q", req.SignatureAlgorithm)
+		}
+
+		if !caps.SupportsAlgorithms(v.Requires...) {
+			return nil, fmt.Errorf("signature algorithm is %q not supported by TPM", req.SignatureAlgorithm)
+		}
```

Comment:
This is additional behavior on top of the new "preferred" algorithm approach; it will verify whether the requested signature algorithm is supported by the TPM.

Reply:
I thought this made sense, but if it's too aggressive of a change for right now, I'll remove.

Reply:
The part checking algorithm support is great, as that'll allow us to return a more descriptive error message, instead of relying on a possibly more cryptic TPM error code. That's been on my wishlist for a while, in fact 😅 We must ensure that it's working as expected with actual TPMs, as it could mean we'd be blocking more operations than before. But I think it's fairly low risk at this time.

If possible, I would make selection of the preferred algorithm something that becomes configurable, as it's fairly opinionated at this time, and thus becomes opt-in. It could be on the `TPMKMS` level when instantiating the KMS, when doing key creation operations, or both.

Reply:
d883afe